File: 76/solution.go
package _76

func minWindow(s string, t string) string {
	n := len(s)
	if n < len(t) {
		return ""
	}
	counter := map[byte]int{}
	i, j, p, q, count := 0, 0, 0, n+1, 0
	for i := 0; i < len(t); i++ {
		c := t[i]
		counter[c]++
		if counter[c] == 1 {
			count++
		}
	}
	for j < n {
		c := s[j]
		j++
		_, ok := counter[c]
		if ok {
			counter[c]--
			if counter[c] == 0 {
				count--
			}
		}
		for i < j && count == 0 {
			if q-p > j-i {
				q = j
				p = i
			}
			c = s[i]
			i++
			_, ok = counter[c]
			if ok {
				if counter[c] == 0 {
					count++
				}
				counter[c]++
			}
		}
	}
	if q == n+1 {
		return ""
	}
	return s[p:q]
}

File: 76/solution.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        n = len(s)
        if n < len(t):
            return ""
        counter = {}
        count = 0
        for c in t:
            if c not in counter:
                count += 1
                counter[c] = 1
            else:
                counter[c] += 1
        i, j, p, q = 0, 0, 0, n
        while j < n:
            c = s[j]
            j += 1
            if c in counter:
                counter[c] -= 1
                if counter[c] == 0:
                    count -= 1
            while i < j and count == 0:
                if q - p + 1 > j - i:
                    p = i
                    q = j - 1
                c = s[i]
                i += 1
                if c in counter:
                    if counter[c] == 0:
                        count += 1
                    counter[c] += 1
        if q == n:
            return ""
        else:
            return s[p:q+1]
